plugins/font/plugin.js

```diff
@@ -88,131 +193,345 @@
-							// Remove it but in a way that we won't lose other empty inline elements inside it.
-							// Example: <p>x<span style="font-size:48px"><em>[]</em></span>x</p>
-							// Result: <p>x<em>[]</em>x</p>
-							if ( startBoundary && endBoundary ) {
-								bm = range.createBookmark();
-								// Replace the element with its children (TODO element.replaceWithChildren).
-								while ( ( node = matching.getFirst() ) ) {
-									node.insertBefore( matching );
-								}
-								matching.remove();
-								range.moveToBookmark( bm );
-
-							// If we are at the boundary of the style element, move out and copy nested styles/elements.
-							} else if ( startBoundary || endBoundary ) {
-								range.moveToPosition( matching, startBoundary ? CKEDITOR.POSITION_BEFORE_START : CKEDITOR.POSITION_AFTER_END );
-								cloneSubtreeIntoRange( range, path.elements.slice(), matching );
-							} else {
-								// Split the element and clone the elements that were in the path
-								// (between the startContainer and the matching element)
-								// into the new place.
-								range.splitElement( matching );
-								range.moveToPosition( matching, CKEDITOR.POSITION_AFTER_END );
-								cloneSubtreeIntoRange( range, path.elements.slice(), matching );
-							}
+			onRender: function() {
+				editor.on( 'selectionChange', function( ev ) {
+					var currentValue = this.getValue(),
+						elementPath = ev.data.path,
+						value = getMatchingValue( editor, elementPath, stylesData );
 
-							editor.getSelection().selectRanges( [ range ] );
+					if ( value ) {
+						if ( value != currentValue ) {
+							this.setValue( value );
 						}
-					} else {
-						editor.removeStyle( previousStyle );
-					}
-				}
 
-				if ( value === this.defaultValue ) {
-					if ( previousStyle ) {
-						editor.removeStyle( previousStyle );
+						return;
 					}
-				} else {
-					editor.applyStyle( style );
-				}
 
-				editor.fire( 'saveSnapshot' );
-			},
+					// If no styles match, just empty it.
+					this.setValue( '', definition.defaultLabel );
+				}, this );
 
-			onRender: function() {
-				editor.on( 'selectionChange', function( ev ) {
-					var currentValue = this.getValue();
-
-					var elementPath = ev.data.path,
-						elements = elementPath.elements;
-
-					// For each element into the elements path.
-					for ( var i = 0, element; i < elements.length; i++ ) {
-						element = elements[ i ];
-
-						// Check if the element is removable by any of
-						// the styles.
-						for ( var value in styles ) {
-							if ( styles[ value ].checkElementMatch( element, true, editor ) ) {
-								if ( value != currentValue )
-									this.setValue( value );
-								return;
-							}
-						}
+				command.on( 'state', function() {
+					if ( command.state === CKEDITOR.TRISTATE_DISABLED ) {
+						this.setState( CKEDITOR.TRISTATE_DISABLED );
+					} else {
+						this.setState( CKEDITOR.TRISTATE_OFF );
 					}
```

Comment:
Do we actually need to check the current command state? :thinking: I see that it won't allow `CKEDITOR.TRISTATE_ON` but if someone really wants to set `ON` state, we should probably allow forcing it. It's the only reason, or there is some additional quirk fixed by checking current command states?

Reply:
> Do we actually need to check the current command state? 

yes

> I see that it won't allow CKEDITOR.TRISTATE_ON but if someone really wants to set ON state, we should probably allow forcing it.

The code doesn't prevent setting `ON` state. You can still use `editor.ui.get( 'Font' ).setState( CKEDITOR.TRISTATE_ON )`

> It's the only reason, or there is some additional quirk fixed by checking current command states?

This is a reason why it is here.
UI doesn't synchronize with the command by itself. This would require to introduce a new feature to rich combo, which is out of the scope for this issue. Probably similar as it was done for the PanelButton for the color commands.

For example, when you would call `editor.getCommand( 'fontSize' ).disable()`, the command will change state to `DISABLED`, however, the richcombo UI won't get this info and remain in `OFF` state. To gave synchronization from command to UI, it is required to listen on command state change. I can extract and describe this case together with information that this part of the font command should be removed. WDYT?
I also don't think, it is a good idea to embed such changes directly in this PR and also I would not recommend making this PR dependent on another one.

Reply:
Sorry I could have said it wrong. I meant we could simplify this code into something like:

```javascript
command.on( 'state', function() {
  this.setState( command.state );
} );
```
so, UI change is totally dependant on command behavior in that case, without additional logic. However, as we already talked F2F we are also setting `CKEDITOR.TRISTATE_ON` within command refresh, so this place should be also simplified:

https://github.com/ckeditor/ckeditor4/blob/b183d83bb103f3bf2e8d6555356ef9b59875205b/plugins/font/plugin.js#L143-L156

Into something like:

```javascript
refresh: function( editor, path ) {
	if ( !defaultContentStyle.checkApplicable( path, editor, editor.activeFilter ) ) {
		this.setState( CKEDITOR.TRISTATE_DISABLED );
	}
}
```